File: days/day_07/solutions/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void solve(char* content, long long* part1_result, long long* part2_result) {
    // Split content into lines
    char** lines = NULL;
    int line_count = 0;
    int capacity = 1000;
    lines = (char**)malloc(capacity * sizeof(char*));
    
    char* pos = content;
    while (*pos != '\0') {
        char* line_start = pos;
        // Find next newline or end of string
        while (*pos != '\n' && *pos != '\0') {
            pos++;
        }
        
        if (line_count >= capacity) {
            capacity *= 2;
            lines = (char**)realloc(lines, capacity * sizeof(char*));
        }
        
        if (*pos == '\n') {
            *pos = '\0';
            pos++;
        }
        
        lines[line_count++] = line_start;
    }
    
    if (line_count == 0) {
        *part1_result = 0;
        *part2_result = 0;
        free(lines);
        return;
    }
    
    // Build grid
    int rows = line_count;
    int cols = strlen(lines[0]);
    char** grid = lines;
    
    // Find starting position S
    int start_row = -1, start_col = -1;
    for (int r = 0; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            if (grid[r][c] == 'S') {
                start_row = r;
                start_col = c;
                break;
            }
        }
        if (start_row != -1) break;
    }
    
    if (start_row == -1) {
        *part1_result = 0;
        *part2_result = 0;
        free(lines);
        return;
    }
    
    // Part 1: Count total splits using sets (represented as arrays)
    int split_count = 0;
    int* active_beams = (int*)calloc(cols, sizeof(int));
    active_beams[start_col] = 1;
    
    for (int r = start_row + 1; r < rows; r++) {
        int* next_beams = (int*)calloc(cols, sizeof(int));
        for (int c = 0; c < cols; c++) {
            if (active_beams[c]) {
                if (grid[r][c] == '.') {
                    next_beams[c] = 1;
                } else if (grid[r][c] == '^') {
                    split_count++;
                    if (c - 1 >= 0) next_beams[c - 1] = 1;
                    if (c + 1 < cols) next_beams[c + 1] = 1;
                }
            }
        }
        free(active_beams);
        active_beams = next_beams;
    }
    free(active_beams);
    
    // Part 2: Count beams reaching bottom row using 2D array
    long long** beam_counts = (long long**)malloc(rows * sizeof(long long*));
    for (int r = 0; r < rows; r++) {
        beam_counts[r] = (long long*)calloc(cols, sizeof(long long));
    }
    beam_counts[start_row][start_col] = 1;
    
    for (int r = start_row + 1; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            long long prev_count = beam_counts[r - 1][c];
            if (prev_count > 0) {
                if (grid[r][c] == '.') {
                    beam_counts[r][c] += prev_count;
                } else if (grid[r][c] == '^') {
                    if (c - 1 >= 0) beam_counts[r][c - 1] += prev_count;
                    if (c + 1 < cols) beam_counts[r][c + 1] += prev_count;
                }
            }
        }
    }
    
    // Sum all beams in bottom row
    long long bottom_beam_count = 0;
    for (int c = 0; c < cols; c++) {
        bottom_beam_count += beam_counts[rows - 1][c];
    }
    
    // Clean up
    for (int r = 0; r < rows; r++) {
        free(beam_counts[r]);
    }
    free(beam_counts);
    
    *part1_result = split_count;
    *part2_result = bottom_beam_count;
    // Note: lines are not freed here as they point to content
}

int main() {
    FILE* file = fopen("../data/input.txt", "r");
    if (file == NULL) {
        fprintf(stderr, "Error opening file\n");
        return 1;
    }
    
    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);
    
    char* content = (char*)malloc(file_size + 1);
    fread(content, 1, file_size, file);
    content[file_size] = '\0';
    fclose(file);
    
    long long part1, part2;
    solve(content, &part1, &part2);
    
    printf("Part 1: %lld\n", part1);
    printf("Part 2: %lld\n", part2);
    
    free(content);
    return 0;
}

File: days/day_07/solutions/solution.clj
(require '[clojure.string :as str])

(defn read-input-raw [path]
  (->> path slurp str/trim))

(defn solve [input-data]
  (let [lines (str/split-lines input-data)
        rows (count lines)
        cols (if (> rows 0) (count (first lines)) 0)
        
        ;; Find starting position S - strings can be indexed directly
        start-pos (first (for [r (range rows)
                               c (range cols)
                               :when (= (get (nth lines r) c) \S)]
                           [r c]))
        start-row (first start-pos)
        start-col (second start-pos)]
    
    (if (nil? start-pos)
      [0 0]
      (let [;; Part 1: Count total splits
            split-count (loop [r (inc start-row)
                               active-beams #{start-col}
                               cnt 0]
                          (if (>= r rows)
                            cnt
                            (let [row (nth lines r)
                                  next-beams (reduce (fn [acc col]
                                                       (case (get row col)
                                                         \. (conj acc col)
                                                         \^ (cond-> acc
                                                              (>= (dec col) 0) (conj (dec col))
                                                              (< (inc col) cols) (conj (inc col)))
                                                         acc))
                                                     #{} active-beams)
                                  new-cnt (+ cnt (count (filter #(= (get row %) \^) active-beams)))]
                              (recur (inc r) next-beams new-cnt))))
            
            ;; Part 2: Count beams reaching bottom row
            beam-counts (reduce (fn [counts r]
                                  (reduce (fn [acc c]
                                            (let [prev-count (get-in acc [(dec r) c])]
                                              (if (> prev-count 0)
                                                (let [row (nth lines r)]
                                                  (case (get row c)
                                                    \. (update-in acc [r c] + prev-count)
                                                    \^ (cond-> acc
                                                         (>= (dec c) 0) (update-in [r (dec c)] + prev-count)
                                                         (< (inc c) cols) (update-in [r (inc c)] + prev-count))
                                                    acc))
                                                acc)))
                                          counts (range cols)))
                                (assoc-in (vec (repeat rows (vec (repeat cols 0)))) [start-row start-col] 1)
                                (range (inc start-row) rows))
            bottom-beam-count (reduce + (get beam-counts (dec rows)))]
        
        [(str split-count) (str bottom-beam-count)]))))

(let [content (read-input-raw "../data/input.txt")
      [part1 part2] (solve content)]
  (println (str "Part 1: " part1))
  (println (str "Part 2: " part2)))
